src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java

```diff
@@ -497,6 +499,18 @@ public RowFilter getPostIndexQueryFilter(RowFilter filter)
         throw new UnsupportedOperationException();
     }
 
+    @Override
+    public void validate(ReadCommand command) throws InvalidRequestException
+    {
+        if (!getIndexContext().isVector())
+            return;
+
+        // to avoid overflow HNSW internal data structure and avoid OOM when filtering top-k
+        if (command.limits().isUnlimited() || command.limits().count() >= MAX_TOP_K)
```

Comment:
What about auto capping to the MAX and adding a WARN that the result set size had a MAX limit applied?